include/arch/arm/aarch64/arm_mmu.h

```diff
@@ -0,0 +1,166 @@
+				(0x44 << (MT_NORMAL_NC * 8))      |	\
+				(0xffUL << (MT_NORMAL * 8)))
+
+/* More flags from user's perpective are supported using remaining bits
+ * of "attrs" field, i.e. attrs[31:3], underlying code will take care
+ * of setting PTE fields correctly.
+ *
+ * current usage of attrs[31:3] is:
+ * attrs[3] : Access Permissions
+ * attrs[4] : Memory access from secure/ns state
+ * attrs[5] : Execute Permissions
+ *
+ */
+#define MT_PERM_SHIFT		3U
+#define MT_SEC_SHIFT		4U
+#define MT_EXECUTE_SHIFT	5U
+
+#define MT_RO			(0U << MT_PERM_SHIFT)
+#define MT_RW			(1U << MT_PERM_SHIFT)
+
+#define MT_SECURE		(0U << MT_SEC_SHIFT)
+#define MT_NS			(1U << MT_SEC_SHIFT)
+
+#define MT_EXECUTE		(0U << MT_EXECUTE_SHIFT)
+#define MT_EXECUTE_NEVER	(1U << MT_EXECUTE_SHIFT)
+
+/* Some compound attributes for most common usages */
+#define MT_CODE			(MT_NORMAL | MT_RO | MT_EXECUTE)
+#define MT_RODATA		(MT_NORMAL | MT_RO | MT_EXECUTE_NEVER)
+
+/*
+ * PTE descriptor can be Block descriptor or Table descriptor
+ * or Page descriptor.
+ */
+#define PTE_DESC_TYPE_MASK	3U
+#define PTE_BLOCK_DESC		1U
+#define PTE_TABLE_DESC		3U
+#define PTE_PAGE_DESC		3U
+#define PTE_INVALID_DESC	0U
+
+/*
+ * Block and Page descriptor attributes fields
+ */
+#define PTE_BLOCK_DESC_MEMTYPE(x)	(x << 2)
+#define PTE_BLOCK_DESC_NS		(1ULL << 5)
+#define PTE_BLOCK_DESC_AP_RO		(1ULL << 7)
+#define PTE_BLOCK_DESC_AP_RW		(0ULL << 7)
+#define PTE_BLOCK_DESC_NON_SHARE	(0ULL << 8)
+#define PTE_BLOCK_DESC_OUTER_SHARE	(2ULL << 8)
+#define PTE_BLOCK_DESC_INNER_SHARE	(3ULL << 8)
+#define PTE_BLOCK_DESC_AF		(1ULL << 10)
+#define PTE_BLOCK_DESC_NG		(1ULL << 11)
+#define PTE_BLOCK_DESC_PXN		(1ULL << 53)
+#define PTE_BLOCK_DESC_UXN		(1ULL << 54)
+
+/*
+ * TCR definitions.
+ */
+#define TCR_EL1_IPS_SHIFT	32U
+#define TCR_EL2_PS_SHIFT	16U
+#define TCR_EL3_PS_SHIFT	16U
+
+#define TCR_T0SZ_SHIFT		0U
+#define TCR_T0SZ(x)		((64 - (x)) << TCR_T0SZ_SHIFT)
+
+#define TCR_IRGN_NC		(0ULL << 8)
+#define TCR_IRGN_WBWA		(1ULL << 8)
+#define TCR_IRGN_WT		(2ULL << 8)
+#define TCR_IRGN_WBNWA		(3ULL << 8)
+#define TCR_IRGN_MASK		(3ULL << 8)
+#define TCR_ORGN_NC		(0ULL << 10)
+#define TCR_ORGN_WBWA		(1ULL << 10)
+#define TCR_ORGN_WT		(2ULL << 10)
+#define TCR_ORGN_WBNWA		(3ULL << 10)
+#define TCR_ORGN_MASK		(3ULL << 10)
+#define TCR_SHARED_NON		(0ULL << 12)
+#define TCR_SHARED_OUTER	(2ULL << 12)
+#define TCR_SHARED_INNER	(3ULL << 12)
+#define TCR_TG0_4K		(0ULL << 14)
+#define TCR_TG0_64K		(1ULL << 14)
+#define TCR_TG0_16K		(2ULL << 14)
+#define TCR_EPD1_DISABLE	(1ULL << 23)
+
+#define TCR_PS_BITS_4GB		0x0ULL
+#define TCR_PS_BITS_64GB	0x1ULL
+#define TCR_PS_BITS_1TB		0x2ULL
+#define TCR_PS_BITS_4TB		0x3ULL
+#define TCR_PS_BITS_16TB	0x4ULL
+#define TCR_PS_BITS_256TB	0x5ULL
+
+#ifndef _ASMLANGUAGE
+/* Region definition data structure */
+struct arm_mmu_region {
+	/* Region Base Physical Address */
+	u64_t base_pa;
+	/* Region Base Virtual Address */
+	u64_t base_va;
+	/* Region size */
+	u64_t size;
```

Comment:
does it make sense to have this as size_t?

Reply:
Had a similar thought earlier. IMO, since this is a highly arch-specific (in particular, AArch64-specific) construct and portability is not of concern, I do not think there would be any benefit in making it `size_t`.

Reply:
Though, it would be a different story if we decide to use `struct arm_mmu_region` for AArch32 as well, but that would be in the future.

Reply:
Keeping it `u64_t` for now.